lib/Runtime/Library/JavascriptPromise.cpp

```diff
@@ -960,6 +975,60 @@ namespace Js
         return TryCallResolveOrRejectHandler(promiseCapability->GetResolve(), handlerResult, scriptContext);
     }
 
+
+    /**
+     * Determine if at the current point in time, the given promise has a path of reactions that result
+     * in an unhandled rejection.  This doesn't account for potential of a rejection handler added later
+     * in time.
+     */
+    bool JavascriptPromise::WillRejectionBeUnhandled()
+    {
+        bool willBeUnhandled = !this->GetIsHandled();
+        if (!willBeUnhandled)
+        {
+            // if this promise is handled, then we need to do a depth-first search over this promise's reject 
+            // reactions. If we find a reaction that 
+            //    - associated promise is "unhandled" (ie, it's never been "then'd")
+            //    - AND its rejection handler is our default "thrower function" 
+            // then this promise results in an unhandled rejection path.
+
+            JsUtil::Stack<JavascriptPromise*, HeapAllocator> stack(&HeapAllocator::Instance);
+            SimpleHashTable<JavascriptPromise*, int, HeapAllocator> visited(&HeapAllocator::Instance);
+            stack.Push(this);
+            visited.Add(this, 1);
+
+            while (!stack.Empty())
+            {
+                JavascriptPromise * curr = stack.Pop();
+                {
+                    JavascriptPromiseReactionList* reactions = curr->GetRejectReactions();
+                    for (int i = 0; i < reactions->Count(); i++)
+                    {
+                        JavascriptPromiseReaction* reaction = reactions->Item(i);
+                        Var promiseVar = reaction->GetCapabilities()->GetPromise();
+
+                        if (JavascriptPromise::Is(promiseVar))
+                        {
+                            JavascriptPromise* p = JavascriptPromise::FromVar(promiseVar);
+                            if (!p->GetIsHandled() && reaction->GetHandler() == p->GetScriptContext()->GetLibrary()->GetThrowerFunction())
```

Comment:
Do you know whether the thrower function is marshalled between contexts correctly to make sure this comparison returns true even if I add `then`s in other contexts?

Reply:
Do you know what the JS looks like to create multiple script contexts?

I spent some time digging through what happens here.   The default "thrower function" in EntryThen, and we get that from the EntryThen's script context - so assuming default `then` method is called on some promise `p`, `p`'s  `scriptContext` + `javascriptLibrary` are the same `EntryThen`s.

Now, it's possible that someone could change the `then` method on a promise to a function from another context, and in that case, things would be off, but I don't know that we care about supporting those cases.

Reply:
Depends on the host; in `ch` I think it's something like `var crossSiteFunc = WScript.LoadScriptFile('someScript.js', 'samethread');`, where you want `someScript.js` to define a function that is then exposed to the original context. In node it's simpler, using `var vm = require('vm'); var ctx = {}; vm.createContext(ctx); vm.runInContext(evalString, ctx)`

Reply:
OK, updated logic to check equality on the underlying FunctionInfo to see if it is the default thrower function.  Also added some test cases where the promise & the then function are from different script contexts.